src/bgw/job_stat.c

```diff
@@ -211,15 +228,40 @@ calculate_next_start_on_failure(TimestampTz finish_time, int consecutive_failure
 	BeginInternalSubTransaction("next start on failure");
 	PG_TRY();
 	{
-		/* ival = retry_period * (consecutive_failures)  */
-		Datum ival = DirectFunctionCall2(interval_mul,
-										 IntervalPGetDatum(&job->fd.retry_period),
-										 Float8GetDatum(multiplier));
-
+		int i;
+		Datum ival;
+		// max wait time to launch job is 1 minute
+		Interval interval_max = { .time = 60000000 };
+		Interval retry_ival = { .time = 2000000 };
+		retry_ival.time += rand_backoff;
+		/* NULL job means the failure was a launch failure */
+		if (job)
+			ival = IntervalPGetDatum(&job->fd.retry_period);
+		else
+		{
+			// random backoff seconds in [2, 2 + 2^f]
+			ival = IntervalPGetDatum(&retry_ival);
+		}
+
+		/* ival = retry_period ^ (consecutive_failures - 1)  */
+		/* arbitrarily choose 2 for exponential growth */
+		if (job)
+		{
+			for (i = 0; i < exponent - 1; i++)
+			{
+				ival = DirectFunctionCall2(interval_mul, ival, Float8GetDatum(2));
+			}
+		}
 		/* ival_max is the ceiling = MAX_INTERVALS_BACKOFF * schedule_interval */
-		Datum ival_max = DirectFunctionCall2(interval_mul,
-											 IntervalPGetDatum(&job->fd.schedule_interval),
-											 Float8GetDatum(MAX_INTERVALS_BACKOFF));
+		Datum ival_max;
+		if (job)
+			ival_max = DirectFunctionCall2(interval_mul,
+										   IntervalPGetDatum(&job->fd.schedule_interval),
+										   Float8GetDatum(MAX_INTERVALS_BACKOFF));
+		else
+		{
+			ival_max = IntervalPGetDatum(&interval_max);
+		}
```

Comment:
you do:
```
		if (job)
		{
		}
		else
		{
		}
```
3 to 4 times here and below. Maybe you should create a meaningful boolean variable for the `if (job)` case, and you should probably only do 1 `if-else` instead of 3 consecutive times to improve readability.